File: CodeWars/Fundamentals/ReverseOrRotate/ReverseOrRotate.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include <assert.h>

bool if_reverse(char* s) {
    long sum = 0;

    for (size_t i = 0 ; i < strlen(s) ; ++i)
        sum += pow(s[i]-'0', 3);

    return sum%2 == 0;
}

char* reverse(char* s) {
    char* result = malloc(strlen(s));
    strcpy(result, "");

    int idx = 0;
    for (int i = strlen(s)-1 ; i >= 0 ; --i)
        result[idx++] = s[i];
    result[idx] = '\0';

    free(s);
    return result;
}

char* rotate(char* s) {
    char* result = malloc(strlen(s));
    strcpy(result, "");

    strcat(result, strdup(s+1));
    size_t len = strlen(result);
    result[len++] = s[0];
    result[len] = '\0';

    free(s);
    return result;
}

char* revrot(char* s, int sz) {
    printf("%s %d \n", s, sz);
    size_t len = strlen(s);

    char* result = malloc(len);
    strcpy(result, "");

    if (sz <= 0 || (size_t)sz > len)
        return result;

    size_t idx = 0;
    while (idx <= len) {
        char* temp = strndup(s+idx, sz);

        if (if_reverse(temp))
            strcat(result, reverse(temp));
        else
            strcat(result, rotate(temp));

        idx += sz;
        if (idx+sz > len)
            break;
    }

    return result;
}

void dotest(char* s, size_t sz, char *expr)
{
    char *act = revrot(s, sz);
    if(strcmp(act, expr) != 0)
        printf("Error. Expected %s but got %s\n", expr, act);
    assert(strcmp(act, expr) == 0);
}

int main() {

    char s[]= "1234";
    dotest(s, 0, "");
    char s1[] = "";
    dotest(s1, 0, "");
    char s2[] = "1234";
    dotest(s2, 5, "");
    char s3[] = "733049910872815764";
    dotest(s3, 5, "330479108928157");
    char s4[] = "73304991087281576455176044327690580265896";
    dotest(s4, 8, "1994033775182780067155464327690480265895");

}

File: CodeWars/Fundamentals/ReverseOrRotate/ReverseOrRotate.rs

fn if_reverse(s: &str) -> bool {
    s.chars().map(|c| (((c as i8)-('0' as i8)) as i64).pow(3)).sum::<i64>()%2 == 0
}

fn rotate(s: &str) -> String {
    let mut result = s[1..].to_string();
    result.push(s.chars().nth(0).unwrap());
    result
}

fn reverse(s: &str) -> String {
    s.chars().rev().collect()
}

fn revrot(s: &str, c: usize) -> String {

    "".to_string()
}

fn testing(s: &str, c: usize, exp: &str) -> () {
    assert_eq!(&revrot(s, c), exp)
}

fn main() {
    testing("1234", 0, "");
    testing("", 0, "");
    testing("1234", 5, "");
    let s = "733049910872815764";
    testing(s, 5, "330479108928157");
    let s = "73304991087281576455176044327690580265896";
    testing(s, 8, "1994033775182780067155464327690480265895");
}
